pkg/codegen/go/utilities.go

```diff
@@ -120,3 +137,47 @@ func enumTitle(s string) string {
 		return "_" + cgstrings.UppercaseFirst(next)
 	})
 }
+
+// Calculate the name of a field in a resource
+func fieldName(pkg *pkgContext, r *schema.Resource, p *schema.Property) string {
+	s := Title(p.Name)
+	var name string
+	if r != nil {
+		name = disambiguatedResourceName(r, pkg)
+	}
+	if !isReservedResourceField(name, s) {
+		return s
+	}
+	var postfix string
+	switch typ := codegen.UnwrapType(p.Type).(type) {
+	case *schema.EnumType:
+		postfix = "Enum"
+	case *schema.ObjectType:
+		postfix = "Type"
+	case *schema.ResourceType:
+		postfix = "Resource"
+	default:
+		if schema.IsPrimitiveType(typ) {
+			switch typ {
+			case schema.StringType:
+				postfix = "Str"
+			case schema.BoolType:
+				postfix = "Bool"
+			case schema.IntType:
+				postfix = "Int"
+			case schema.NumberType:
+				postfix = "Num"
+			}
+		}
+	}
```

Comment:
Could we simplify this to adding a `_` suffix? Each object type can only have a single field named `ElementType` for example, I don't see what we gain by adding type info in the field name.

Reply:
We could, I just thought that this looked more natural. Its inline with how we deal with `*schema.Object` and `*schema.Resource` name conflicts.